### PR TITLE
PostgreSQL Support

### PR DESCRIPTION
**Pull request by Qstick**
#### Database Migration
YES - 188 (Timezone adjustments to DateTime column types)

#### Description
This adds Postgres Support via config file parameters. Based on Radarr etc.. implementation

#### Todos
- [ ] PG Tests on TC if wanted
- [ ] Wiki Updates

### CODE REVIEW

**Review by markus101 — CHANGES_REQUESTED** [review]
Untested, mostly would just like to see some consistency in the SQL strings.

File: src/NzbDrone.Core/Datastore/Database.cs
Comment (Qstick) [reply src/NzbDrone.Core/Datastore/Database.cs]: Already reworked

File: src/NzbDrone.Core/Datastore/Database.cs
Comment (Qstick) [reply src/NzbDrone.Core/Datastore/Database.cs]: Already Reworked

File: src/NzbDrone.Core/Datastore/Migration/009_fix_renameEpisodes.cs
Comment (Qstick) [reply src/NzbDrone.Core/Datastore/Migration/009_fix_renameEpisodes.cs:13]: Correct

File: src/NzbDrone.Core/Datastore/Migration/018_remove_duplicates.cs
Comment (Qstick) [reply src/NzbDrone.Core/Datastore/Migration/018_remove_duplicates.cs:38]: Yea, this all depends on how the string is concatenated, it's done differently in different places. I'll see about consolidating.

File: src/NzbDrone.Core/Datastore/DbFactory.cs
Comment (Qstick) [reply src/NzbDrone.Core/Datastore/DbFactory.cs:134]: I can move the sleep up to delay the initial retry, but we'll need the `if` below regardless to determine to loop again or throw.

File: src/NzbDrone.Core/Datastore/Migration/018_remove_duplicates.cs
Comment (Qstick) [reply src/NzbDrone.Core/Datastore/Migration/018_remove_duplicates.cs:38]: Essentially this comes down to if strings are created with `string.format()`, `""` or `$""` then `\"` is used. If the string is concatenated with `@""` then `""` is required for escape.

Comment (markus101) [reply src/NzbDrone.Core/Datastore/Migration/018_remove_duplicates.cs:38]: Yeah, at this point we should move away from `string.Format` in general, which means we'd most often use `$`, but right now we sometimes use that or sometimes `@` for these SQL strings. We could use `$@` (or is it `@$`?) if we want to use `""` instead of `\"`, but really just looking for it to be consistent so we can keep it consistent in the future.

Comment (Qstick) [reply src/NzbDrone.Core/Datastore/Migration/018_remove_duplicates.cs:38]: We'd need to move normal string literals as well if we want to use `""`, a string defined with just `"string"` uses `\"` for escape. Seems like verbatim strings (`@""`) and `""` escape are the outliers really. Are there cases where we have to use verbatim for SQL strings?

Comment (markus101) [reply src/NzbDrone.Core/Datastore/Migration/018_remove_duplicates.cs:38]: I don't think I saw any when I was reviewing.

**Comment by markus101** [comment]
Fixing that shouldn't be contingent on landing this larger, unrelated PR though.

**Comment by Qstick** [comment]
Rebased and comments addressed

**Comment by buroa** [comment]
👀

**Comment by martimarkov** [comment]
Is there anything blocking this PR going into the develop branch?

**Comment by Emalton** [comment]
This would be a great addition for those of us using Kubernetes!

**Comment by politeauthority** [comment]
I can't contribute here as it's out of my knowledge space. However I'm working on k8s implementations of the servarr stack, and this contribution is paramount (to my current knowledge) to making this work.
Exited to see this development!

**Comment by onedr0p** [comment]
@politeauthority As someone who's been using Kubernetes for the past 5 years and knows it pretty well, I don't see how this feature would be _paramount_ to get it working in Kubernetes. Sonarr & SQLite works just fine with block storage like rook-ceph, openebs or longhorn and has been working fine like for for years for me.

This is just a _nice to have_ in my opinion because you can treat the configuration volume as ephemeral to a point. If the config volume is lost and use have the config.xml mounted as a secret (or do `envsubt` across it) you can just have Sonarr pull down the metadata again. It's not like you can scale Sonarr if this feature lands, it's a monolithic app.

**Comment by patrickjmcd** [comment]
@onedr0p humbly, I disagree. rook, ceph, and longhorn are all excellent products, but are a jackhammer when looking for a screwdriver. 

Many of us use a nfs provider for config for many services, but when services **depend heavily** on any database, it's really nice to be able to build (and manage) a Postgres database and have your applications use it. Having an advanced storage system (a la ceph, rook, etc) is an advanced k8s concept where _many_ people are following tutorials that lead them to even more simplistic NFS connections. 

I totally understand the position of "we built this to use SQLite and the community offers lots of options for HA even with SQLite", but the real HA truly is a PostgresDB-compliant interface where that db can be self-hosted as well. 

I hope this isn't coming off as attacking or rude, but as someone who's in the "more advanced at k8s" crew, but looking for the simplest solution for everything, I think that this is **super** valuable to the community.

**Comment by patrickjmcd** [comment]
And not to pile on more, but @Qstick made it totally optional to take this path, but made it available to anyone who wants to use it. What's the harm in that?

**Comment by onedr0p** [comment]
I never said it wasn't valuable and in fact I already use Postgres for Radarr and Prowlarr. I would use it since I already have cloudnative-pg up and running in my cluster. I just don't think there's a need to keep commenting on this PR asking for a status or to show some sort of urgency that it's needed _now_.

**Comment by patrickjmcd** [comment]
Ah ok, I'm sorry @onedr0p! Just trying to put my $0.02 behind this PR

**Review by markus101 — APPROVED** [review]
Thanks for all the fixes and CI is happy (and ready to test future builds). Do we need to do anything else before landing?

**Comment by Qstick** [comment]
Yea, let me rebase on latest just to ensure there are no lingering queries or migrations that need to be fixed.

**Comment by varac** [comment]
@Qstick Please rebase and push, would be great to have this merged soon. Thanks for this PR!

**Comment by Qstick** [comment]
Please don't comment with spam that adds no value to the PR.

**Comment by Qstick** [comment]
There's an issue here with Episode Missing Search and Episode Cutoff unmet search that results in an ambiguous order by column (Id) error when making the call to PG, SQLite just seems to make some assumptions and go with it. This needs sorting before this goes in.

**Comment by kharenis** [comment]
@Qstick Hi, I've created a PR that resolves the Epsisode Missing Search and Episode Cutoff unmet search issues.

**Comment by nwithan8** [comment]
For those that aren't super technical, can we get a 30,000-foot overview of the performance improvements we can expect with this change to the backend database? Radarr already uses Postgres and I'm unable to immediately tell a difference between Sonarr and Radarr performance...

**Comment by VonLatvala** [comment]
> For those that aren't super technical, can we get a 30,000-foot overview of the performance improvements we can expect with this change to the backend database? Radarr already uses Postgres and I'm unable to immediately tell a difference between Sonarr and Radarr performance...

For example, running this using sqlite on a container host having its PVCs (persistent volumes) mounted over NFS, the database is locked to an almost unusable level. Using postgres over the network works like a breeze. This is just from on the top of my head, someone else might have more rigorous numbers. Cheers!

**Comment by Qstick** [comment]
> @Qstick Hi, I've created a PR that resolves the Episode Missing Search and Episode Cutoff unmet search issues.

The table we are querying is not necessarily the sort key table... thus this fix causes other issues.

**Comment by Qstick** [comment]
@markus101 Believe we are ready, good with this pushing in after a squash now?

**Comment by Qstick** [comment]
> For those that aren't super technical, can we get a 30,000-foot overview of the performance improvements we can expect with this change to the backend database? Radarr already uses Postgres and I'm unable to immediately tell a difference between Sonarr and Radarr performance...

99% of Sonarr or Radarr datasets are not what most people would consider huge datasets. Postgres gains a tad and obviously has other advantages, but it’s no magic speed bullet. How we query and how those queries are setup have much more impact to speed, and we’ve done a lot of work around that in Sonarr v4. Radarr uses Sqlite by default, you have to manually configure postgres via your config file in either app.

**Comment by markus101** [comment]
@Qstick looks like CI is failing on a number of tests due to migration 192.

**Comment by Qstick** [comment]
Thanks, I looked at the CI a bit to quick, it hadn't finished last build.

I'll fix

**Comment by Qstick** [comment]
Should be better now, Although I don't like that fluentmigrator doesn't handle that PG column alter cleanly

**Comment by markus101** [comment]
Yeah, that's annoying, looks good now though, squash it up and ship it!

**Comment by martimarkov** [comment]
I'm sorry for the spam but amazing job!!! I've been waiting for this feature for years now!!!

Thanks @Qstick ❤️❤️❤️

**Comment by buroa** [comment]
Thanks so much for this @Qstick; https://github.com/buroa/k8s-gitops/commit/005efd5dc338ab6a7548fc1c186a868f8b1609c8 :)

**Pull request merged**
